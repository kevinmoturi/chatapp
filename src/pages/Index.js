import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';

const Index = () => {
    const navigate = useNavigate();
    const [name, setName] = useState('');
    return (
        <div className='flex flex-col justify-center items-center w-full h-screen bg-gray-900'>
            <div className='max-w-md bg-white rounded w-full p-10'>
                <h1 className="text-3xl font-bold text-gray-700 text-center">Welcome to Chat Room</h1>
                <p className='text-sm text-gray-500 capitalize text-center'>Name field is required</p>
                <form onSubmit={(e) => {
                    e.preventDefault();
                    navigate('/chatroom');
                }}>
                    <div className='my-5'>
                        <label for="name" className='font-semibold text-gray-500'>Name (s)</label>
                        <input id="name" name="name" type="text" autoComplete='off' required className="rounded relative block w-full px-3 py-2 border border-gray-300 placeholder-gray-500 text-gray-900 focus:outline-none focus:ring-gray-500 focus:border-gray-500 focus:z-10 sm:text-sm mt-2" placeholder="e.g John" value={name} onChange={e => setName(e.target.value)} />
                    </div>
                    <div className='my-8'>
                        <button type='submit' className='w-full bg-gray-900 text-gray-200 py-2 rounded shadow'>Enter Chat Room</button>
                    </div>
                </form>
            </div>
        </div>
    )
};

export default Index;
