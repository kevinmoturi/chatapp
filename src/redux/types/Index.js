export const SET_USER = "SET_USER";
export const GET_USER = "GET_USER";

export const GET_CHATS = 'GET_CHATS';
export const NEW_CHAT = 'NEW_CHAT';