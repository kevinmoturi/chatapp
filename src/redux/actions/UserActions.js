import { GET_USER, SET_USER } from './../types/Index';

export const getUser = () => async dispatch => {

    dispatch({ type: GET_USER });

    return null;

};


export const setUser = (name) => async dispatch => {

    dispatch({ type: SET_USER, payload: name });

    return null;

};