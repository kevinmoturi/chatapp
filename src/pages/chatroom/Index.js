import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { getUser } from './../../redux/actions/UserActions';
import { getChats, sendMessage } from './../../redux/actions/ChatActions';

const Index = ({ getUser: getCurrentUser, user, getChats: getMessages, chats, sendMessage: postMessage }) => {
    const [message, setMessage] = useState('');
    useEffect(() => {
        getCurrentUser();
    }, [getCurrentUser, getMessages]);
    useEffect(() => {
        getMessages();
    });
    const scrollToBottom = () => {
        let el = document.getElementById('messages');
        let scroll_height = el.scrollHeight, client_height = el.clientHeight;
        let maxScrollTop = scroll_height - client_height;
        el.scroll({ top: maxScrollTop > 0 ? maxScrollTop : 0, behavior: 'smooth' });
    }
    if(!user) return window.location.href = '/';
    return (
        <div className='flex flex-col justify-center items-center w-full h-screen bg-white'>
            <div className='max-w-md bg-white rounded w-full flex flex-col'>
                <div className='bg-gray-700 p-5 rounded-t'>
                    <h1 className='text-white'>Chat Room</h1>
                </div>
                <div className='h-72 bg-slate-200 overflow-y-auto pb-5' id='messages'>
                    {chats.map((chat, i) => {
                        if(chat.sender === user) {
                            return <div key={i} className='flex flex-row justify-end items-start gap-x-3 mt-5 px-2'>
                                <p className='bg-gray-600 text-white py-2 px-6 rounded-sm'>{chat.message}</p>
                                <img className='bg-gray-400 h-10 rounded-full object-contain' src='/img/profile.png' alt='avatar' />
                            </div>
                        } else {
                            return <div key={i} className='flex flex-row justify-start items-start gap-x-3 mt-5 px-2 w-full'>
                                <img className='bg-gray-400 h-10 rounded-full object-contain' src='/img/profile.png' alt='avatar' />
                                <p className='bg-gray-600 text-white py-2 px-6 rounded-sm'>{chat.message} - {chat.sender}</p>
                            </div>
                        }
                    })}
                </div>
                <form className='flex flex-row items-center bg-gray-700 p-5 rounded-b' onSubmit={e => {
                    e.preventDefault();
                    let obj = { message, sender: user };
                    postMessage(obj);
                    setMessage('');
                    scrollToBottom();
                }}>
                    <input id="message" name="message" type="text" autoComplete='off' required className="rounded-full relative block w-full px-5 py-2 border border-gray-300 placeholder-gray-500 text-gray-900 focus:outline-none focus:ring-gray-500 focus:border-gray-500 focus:z-10 sm:text-sm" placeholder="Write ..." value={message} onChange={e => setMessage(e.target.value)} />
                    <button type='submit' className='bg-white text-gray-700 uppercase py-2 px-5 ml-3 rounded-full'>Send</button>
                </form>
            </div>
        </div>
    )
};

const mapStateToProps = (state, ownProps) => {
    return {
        user: state.user,
        chats: state.chats
    }
};

export default connect(mapStateToProps, { getUser, getChats, sendMessage })(Index);
