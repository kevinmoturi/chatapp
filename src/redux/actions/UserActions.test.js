import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getUser, setUser } from './UserActions';
import { GET_USER, SET_USER } from './../types/Index';

// Create a mock store
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({});

describe('user actions', () => {
    it('action type should be GET_USER and snapshot taken', () => {
        store.dispatch(getUser());
        expect(store.getActions()).toMatchSnapshot();
        expect(store.getActions()[0].type).toEqual(GET_USER);
    });
    it('action type should be SET_USER and should pass name in the payload and snapshot taken', () => {
        let name = 'Kevin';
        store.dispatch(setUser(name));
        expect(store.getActions()).toMatchSnapshot();
        expect(store.getActions()[1].type).toEqual(SET_USER);
        expect(store.getActions()[1].payload).toEqual(name);
    });
});