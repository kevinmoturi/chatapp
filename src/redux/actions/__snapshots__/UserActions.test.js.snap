// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user actions action type should be GET_USER and snapshot taken 1`] = `
Array [
  Object {
    "type": "GET_USER",
  },
]
`;

exports[`user actions action type should be SET_USER and should pass name in the payload and snapshot taken 1`] = `
Array [
  Object {
    "type": "GET_USER",
  },
  Object {
    "payload": "Kevin",
    "type": "SET_USER",
  },
]
`;
